### ISSUE TITLE
feature request: hidden partitioning / timestamp partitioning

### ISSUE THREAD

**Issue by changpingc** [delta-io/delta#490]
We store a lot of time series data using delta format. They all have a timestamp column and are partitioned by `date(timestamp)`. 

Two common mistakes I see are
(1) forgetting to include a date filter in addition to timestamp filter
(2) partitioning output by timestamp leading to millions of output folders

The problem goes away if delta supports something like hidden partitioning in Iceberg, or specifying timestamp partition granularity (daily vs hourly) in BigQuery.

https://iceberg.apache.org/partitioning/#icebergs-hidden-partitioning
https://cloud.google.com/bigquery/docs/creating-column-partitions#daily_partitioning_vs_hourly_partitioning

**Comment by brkyvz** [comment]
Yes, this would be a fantastic addition to Delta!

**Comment by gwax** [comment]
This would be very useful for geospatial data indexing with [H3](https://h3geo.org/). If it were possible to use a truncated (lower resolution) index as a partition but query against the full index without having to manually truncate we could accelerate geospatial queries while hiding partition details under the hood.

**Comment by rubenssoto** [comment]
Is there any plans to support this in delta?

**Comment by brkyvz** [comment]
yes, work is starting this week! Will be a long process though - it's a big project to get right

**Comment by Tagar** [comment]
@gwax what was the idea with geospatial H3 indexing indexes through hidden partitions? Can you share more details? 

Generated Columns (that can be used for partitioning ) were added recently 
https://docs.databricks.com/delta/delta-batch.html#deltausegeneratedcolumns

**Comment by MasterDDT** [comment]
We have a slightly different use case, our data is partitioned by a userId (e.x. long). So of course we dont want `partitionBy` (one file per-partition-value). And we dont want to use `bucketBy` because it requires a metastore.

Is it possible to just read the `SparkPlan` of the dataframe and write out the partitioning/sorting information to `_delta_log/`? So we can `df.repartition(5, col1).write.format("delta").save(s3://123)` and then `read.format("delta").load(s3://123)` and the data is already hash partitioned by `col1` with 5 partitions? Maybe only support it for hash partitioning with the default mod function, not for custom-functions.

**Comment by keen85** [comment]
@brkyvz are you still working on this topic?

**Comment by dishkakrauch** [comment]
Any changes to see updates here?

**Comment by souzomain** [comment]
hello, any work arround this?

**Comment by foss-contributor** [comment]
@changpingc, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!

**Comment by souzomain** [comment]
hey @brkyvz can I help you with this? I need this feature for my use case
